www/js/lib/popovers.js

```diff
@@ -0,0 +1,455 @@
+                                // embedded link, so only use innerText in that case
+                                const content = params.contentInjectionMode === 'jquery' ? plainText
+                                    : nonEmptyParagraphs[i].innerHTML;
+                                balloonString += '<p>' + content + '</p>';
+                                // If we have enough characters to fill the box, break
+                                if (cumulativeCharCount >= 850) break;
+                            }
+                        }
+                        // If we have a lede, we can now add an image to the balloon
+                        const images = articleBody.querySelectorAll('img');
+                        let firstImage = null;
+                        if (images && params.contentInjectionMode === 'serviceworker') {
+                            // Iterate over images until we find one with a width greater than 50 pixels
+                            // (this filters out small icons)
+                            const imageArray = Array.from(images);
+                            for (let j = 0; j < imageArray.length; j++) {
+                                if (imageArray[j] && imageArray[j].width > 50) {
+                                    firstImage = imageArray[j];
+                                    break;
+                                }
+                            }
+                        }
+                        if (firstImage) {
+                            // Calculate root relative URL of image
+                            const imageZimURL = encodeURI(uiUtil.deriveZimUrlFromRelativeUrl(firstImage.getAttribute('src'), balloonBaseURL));
+                            firstImage.src = rootRelativePathPrefix + imageZimURL;
+                            balloonString = firstImage.outerHTML + balloonString;
+                        }
+                        if (!balloonString) {
+                            reject(new Error('No article lede or image'));
+                        } else {
+                            resolve(balloonString);
+                        }
+                    } else {
+                        reject(new Error('No article body found'));
+                    }
+                });
+            });
+        }
+        if (!dirEntry) {
+            return Promise.reject(new Error('No directory entry found'));
+        } else if (dirEntry.redirect) {
+            // If the dirEntry is a redirect, we need to resolve it before reading the article
+            return new Promise((resolve, reject) => {
+                archive.resolveRedirect(dirEntry, function (reDirEntry) {
+                    resolve(readArticle(reDirEntry));
+                });
+            }).catch(error => {
+                return Promise.reject(error);
+            });
+        } else {
+            // Directory entry was found, so now read the article data
+            return Promise.resolve(readArticle(dirEntry));
```

Comment:
Again, might be rusty, but you're already in a function that is called from a `then()` right? So you shouldn't need `Promise.resolve()`, you should just be able to `return readArticle(dirEntry)`?